lib/url.js

```diff
@@ -516,6 +516,13 @@ function getHostname(self, rest, hostname) {
                      code !== CHAR_COLON);
 
     if (!isValid) {
+      // If leftover starts with :, then it represents an invalid port.
+      // But url.parse() is lenient about it for now.
+      // Issue a warning and continue.
+      if (code === CHAR_COLON) {
+        const detail = `The URL ${url} is invalid. Future versions of Node.js will throw an error.`;
+        process.emitWarning(detail, 'DeprecationWarning', 'DEP0170');
```

Comment:
It might be worthwhile to limit the warning to once-per-process to avoid spamming logs.

Reply:
As written would it be once per url?

Reply:
> It might be worthwhile to limit the warning to once-per-process to avoid spamming logs.

I think it would be better to warn on each invalid URL but if you feel strongly that it should be once per process, I'll change it. I'm interpreting your comment as a gentle suggestion and not a strong opinion.

Reply:
Strong suggestion ;-) I've seen first hand cases of url.parses that could trigger this warning on every request to a server. Given how emitWarning works with nextTick, that could end up generating a fairly significant performance hit along with just being spammy in general. But, it's also a great motivation to change APIs.

Reply:
> Strong suggestion ;-)

OK, @jasnell: Strong suggestion, incorporated. PTAL.